### PR TITLE
Upgrade to Tokio 1.0

### LINKED ISSUE #725: Upgrade to Tokio v1
**Dependencies which need to upgrade**
- [x] hyper (hyperium/hyper#2369)
- [x] tokio-rustls (tokio-rs/tls#46)
- [x] tokio-tungstenite (snapview/tokio-tungstenite#142)

### PR DESCRIPTION
**Pull request by aknuds1**
Similar to @Urhengulas' PR #741, although more complete, this PR aims to upgrade to Tokio v1.0.

I can successfully run all tests with all features enabled: `cargo test --all-features`.

This is my first time trying to contribute to warp, and I'm not super familiar with the internals, so please consider it a best effort :) 

Fixes #725.

### CODE REVIEW

File: Cargo.toml
Comment (Urhengulas) [inline Cargo.toml:20]: Do you really need the `rev`, or is it just pointing to master?

File: src/filters/fs.rs
Comment (Urhengulas) [inline src/filters/fs.rs:540]: Maybe add `Safety: `dst` may include uninitialized memory, but `ReadBuf` will never uninitialize memory that has already been initialized.` as comment? (See https://github.com/tokio-rs/tokio/discussions/3121)

**Review by Urhengulas — COMMENTED** [review]
Good job! Only two remarks :D

File: Cargo.toml
Comment (aknuds1) [reply Cargo.toml:20]: It's more the other way around, master is pointing to this revision. At least it was, which is why I prefer pinning the revision, since a branch by its nature will change and you have no guarantee it will keep working.

The intent is also to switch to official versions of dependencies, before finalizing this PR. The current dependencies on Git commits just represent what I could make work.

File: src/filters/fs.rs
Comment (aknuds1) [reply src/filters/fs.rs:540]: I just copied this code from tokio-util since I couldn't get the latter to work as a dependency. I'm more or less assuming that tokio-util will be used instead by the time this PR is finalized, although I don't know what the maintainers prefer.

File: src/filters/ws.rs
Comment (fundon) [inline src/filters/ws.rs:137]: We can merge
`Response::new(self.ws.body)` and `let mut res = http::Response::default();`  to

```
let mut res   = Response::new(self.ws.body); 
let fut = hyper::upgrade::on(&mut res)
```

Comment (aknuds1) [reply src/filters/ws.rs:137]: @seanmonstar Is it safe to replace `res` as suggested in the above comment? AFAICT, the difference is that `res` gets constructed from `self.ws.body`.

File: Cargo.toml
Comment (jxs) [inline Cargo.toml:39]: `tokio-tungstenite` has released version `0.12` using tokio 0.3

Comment (aknuds1) [reply Cargo.toml:39]: @jxs Great, will update.

Comment (aknuds1) [reply Cargo.toml]: Done.

File: src/filters/fs.rs
Comment (jxs) [reply src/filters/fs.rs:540]: `tokio-util` has since been upgraded to `0.5.1` which has `poll_read_buf`. I think it makes sense to use it and isolate the unsafe code there, but then maybe the `fs` filter should be feature gated as it now introduces an extra dependency.
What do you think @seanmonstar ?

Comment (aknuds1) [reply src/filters/fs.rs:540]: Thanks for the update @jxs :)

**Comment by jhpratt** [comment]
Tokio 1.0 has been released, so this PR should probably be updated to depend on that.

**Comment by aknuds1** [comment]
I'm trying to upgrade to Tokio 1.0.

**Comment by aknuds1** [comment]
The remaining issue is tokio-tungstenite. Trying to upgrade it to tokio 1.0, but it's easier said than done.

**Comment by aknuds1** [comment]
Forked tokio-tungstenite, to upgrade it to tokio 1, and depending on it, but build still fails.

**Comment by paolobarbolini** [comment]
> but build still fails

* `Sleep` is now `!Unpin`
* `bytes_mut` is now `chunk_mut`
* `bytes::buf::{Buf, BufMut}` -> `bytes::{Buf, BufMut}` (it's top-level now)

**Comment by aknuds1** [comment]
If anyone has tips wrt. fixing the build failures, I'm all ears :)

File: Cargo.toml
Comment (paolobarbolini) [inline Cargo.toml:20]: This can now be upgraded to async-compression 0.3.7

Comment (aknuds1) [reply Cargo.toml:20]: @paolobarbolini Looks as if that version still depends on bytes 0.5 if you use the stream feature?

Comment (aknuds1) [reply Cargo.toml:20]: I tried building against it and couldn't make it work.

Comment (paolobarbolini) [reply Cargo.toml:20]: See https://github.com/seanmonstar/reqwest/pull/1076#discussion_r548536489

What you can do is something like [this](https://github.com/seanmonstar/reqwest/pull/1076/files#diff-668dbc2595c2151fbc2dde0df240f68963eb8e30582b5cbc71cf80da947d2800), where you wrap the encoder into `tokio_util::io::StreamReader` and `tokio_util::codec::FramedRead`

File: src/filters/body.rs
Comment (paolobarbolini) [inline src/filters/body.rs:235]: `Buf::chunk` may not return the entire buffer (see https://docs.rs/bytes/1.0.0/bytes/trait.Buf.html#tymethod.chunk)

```suggestion
        serde_json::from_slice(&buf.copy_to_bytes(buf.remaining())).map_err(Into::into)
```

File: Cargo.toml
Comment (aknuds1) [reply Cargo.toml:20]: @paolobarbolini OK, do you have any tips for the remaining failures?

Comment (paolobarbolini) [reply Cargo.toml:20]: [`alive_timer`](https://github.com/seanmonstar/warp/blob/5d068ae90f65cf66784c62daba8a27eaa506fa26/src/filters/sse.rs#L487) needs to be pinned, just like `event_stream` is (add `#[pin]` on top of it).

`Sleep::reset` doesn't take `self` anymore ([before](https://docs.rs/tokio/0.3.6/tokio/time/struct.Sleep.html#method.reset) - [after](https://docs.rs/tokio/1.0.0/tokio/time/struct.Sleep.html#method.reset)), so the two calls to `reset` need to be changed like so:
`first_argument.reset(second_argument)` -> `Sleep::reset(first_argument, second_argument)`

Comment (seanmonstar) [reply Cargo.toml:20]: You can still call it like a method, you just might need an `timer.as_mut().reset(arg)` in there.

Comment (aknuds1) [reply Cargo.toml:20]: Thanks, I fixed remaining build failures. I can look at async-compression next.

File: src/test.rs
Comment (paolobarbolini) [inline src/test.rs:94]: This broke compiling without default features.

```suggestion
```

**Comment by aknuds1** [comment]
@paolobarbolini @seanmonstar I could build with `--all-features` now, but some examples fail when I run tests.

**Comment by aknuds1** [comment]
Is anyone able to help with the remaining build failure?

File: src/filters/ws.rs
Comment (fundon) [reply src/filters/ws.rs:137]: I ported `ws.rs` to viz-rs. 
You can see that https://github.com/viz-rs/viz/blob/master/viz-core/src/ws.rs.

**Comment by ebkalderon** [comment]
Seems like `cargo fmt` is complaining on the most recent commit. Doesn't seem to be a build failure.

File: src/filters/ws.rs
Comment (fundon) [reply src/filters/ws.rs:137]: https://github.com/hyperium/hyper/blob/master/src/upgrade.rs#L258-L287

**Comment by aknuds1** [comment]
With the kind help of @paolobarbolini I could fix all the build failures. Now some test failures remain AFAICT.

**Comment by aknuds1** [comment]
Is anyone able to see why e.g. the integration test `echo_pings` fails?

**Comment by fundon** [comment]
@ aknuds1  
https://github.com/hyperium/hyper/blob/master/src/upgrade.rs#L258-L287

```rust
hyper::upgrade::on(Response::new(self.ws.body));
```
The `Response::new(self.ws.body)` need include the `hyper::upgrade::OnUpgrade` from `req.extensions`, or `hyper::upgrade::on(req);`.

**Comment by aknuds1** [comment]
@fundon Do you have a patch I can apply?

**Comment by fundon** [comment]
@aknuds1  I ported `ws.rs` to https://github.com/viz-rs/viz/blob/master/viz-core/src/ws.rs. You can get it.

**Comment by aknuds1** [comment]
Thanks @fundon! With the help of your code, I could figure out how to handle websocket upgrades.

**Comment by aknuds1** [comment]
All the tests pass, it looks as if tokio/uds needs to be removed as a required check in the project settings if I'm not mistaken (since it's removed from tokio and .github/workflows/ci.yml).

File: Cargo.toml
Comment (Urhengulas) [inline Cargo.toml:44]: Maybe move `async-stream` to be 2nd dependency to keep alphabetical ordering?

**Review by Urhengulas — COMMENTED** [review]
Good work :)

**Comment by hasezoey** [comment]
@aknuds1 could you update this PR? because currently this dosnt build anymore:
```
    Updating git repository `https://github.com/nickelc/tungstenite-rs.git`
error: failed to get `tungstenite` as a dependency of package `tokio-tungstenite v0.12.0 (https://github.com/aknuds1/tokio-tungstenite.git?rev=cad06f3ac07e7fc2a936308207aab72b31478462#cad06f3a)`
    ... which is depended on by `warp v0.2.5 (https://github.com/aknuds1/warp?branch=chore/upgrade-tokio#f6508fa1)`
    ... which is depended on by `myproject v0.1.0 (/myproject)`

Caused by:
  failed to load source for dependency `tungstenite`

Caused by:
  Unable to update https://github.com/nickelc/tungstenite-rs.git?branch=bytes

Caused by:
  failed to find branch `bytes`

Caused by:
  cannot locate remote-tracking branch 'origin/bytes'; class=Reference (4); code=NotFound (-3)

```

**Comment by aknuds1** [comment]
Thanks for the heads up @hasezoey, fixed.

**Comment by apexys** [comment]
Is there anything we could do to help with this PR? 
We would really like to update our application to tokio 1.0 to use some of the newer features, but depend on warp.

**Comment by aknuds1** [comment]
> Is there anything we could do to help with this PR?
> We would really like to update our application to tokio 1.0 to use some of the newer features, but depend on warp.

@apexys I'm waiting on tokio-tungstenite to release a new version, so I don't have to depend on a Git revision.

**Comment by aknuds1** [comment]
I changed the dependency to the latest release of tokio-tungstenite now, i.e. 0.13, so the PR should be ready to merge.

**Comment by mraerino** [comment]
@seanmonstar there seems to be a challenge around the required CI checks since one of them was renamed - you might need to remove the requirement for this to merge and add the new required check after merge

File: src/filters/compression.rs
Comment (jxs) [inline src/filters/compression.rs:61]: why did you prefer `FramedRead` over `ReaderStream`?

Comment (aknuds1) [reply src/filters/compression.rs:61]: @jxs To be honest, I don't remember exactly, but it was probably from an example. Do you think ReaderStream is preferable?

Comment (paolobarbolini) [reply src/filters/compression.rs:61]: It was inspired from the reqwest implementation https://github.com/seanmonstar/reqwest/blob/a19eb3419637856d6d42f55dff6e4790613b2933/src/async_impl/decoder.rs#L47

Comment (aknuds1) [reply src/filters/compression.rs:61]: Thanks @paolobarbolini, I can't bring myself to remember where I found this solution.

Comment (jxs) [reply src/filters/compression.rs:61]: thanks @paolobarbolini. @aknuds1 I asked because I was not sure either, Previously `FrameReader` was the way to achieve `AsyncRead` to `Stream` conversion, but since `ReaderStream` [was introduced](https://github.com/tokio-rs/tokio/issues/2705) I would maybe suggest `ReaderStream` as it's simpler, and recommended for this cases on [tokio-util doc](https://docs.rs/tokio-stream/0.1.1/tokio_stream/#conversion-to-and-from-asyncreadasyncwrite)

**Review by jxs — APPROVED** [review]
LGTM thanks! :fireworks: 
- `Test (uds, --features tokio/uds)` also needs to be renamed on the repository settings
- should the `tokio_util` dependent filters (`fs`, `compression`) also be featured gated?

File: src/filters/compression.rs
Comment (aknuds1) [reply src/filters/compression.rs:61]: @seanmonstar Do you have any opinion on this?

Comment (aknuds1) [reply src/filters/compression.rs]: I converted from FramedRead into ReaderStream, ass @jxs recommended. I don't really know personally if there are any practical differences, but I can tell that the API is simpler (a second argument to `FramedRead::new` can be dropped). I trust @jxs knows more about this than I do, and I'll let @seanmonstar be the judge.

File: Cargo.toml
Comment (seanmonstar) [inline Cargo.toml:25]: Do we need the `client` feature?

File: src/filters/ws.rs
Comment (seanmonstar) [inline src/filters/ws.rs:145]: If you already have the extension, you don't need to insert it here to remove it again. (`on_upgrade` is the same as `fut`).

**Review by seanmonstar — CHANGES_REQUESTED** [review]
Excellent work! I believe this is very close.

File: .github/workflows/ci.yml
Comment (aknuds1) [reply .github/workflows/ci.yml:48]: @seanmonstar OK, I can remove it.

File: Cargo.toml
Comment (aknuds1) [reply Cargo.toml:25]: @seanmonstar Tested now. It's required for the tests. If I add hyper to dev-dependencies though, with only the client feature enabled, they work. I'll make that change, let me know what you think.

File: src/filters/ws.rs
Comment (aknuds1) [reply src/filters/ws.rs]: @seanmonstar I think I see what you mean, and try to simplify it to the best of my ability. Please have a look.

**Comment by koivunej** [comment]
Should the return type of `warp::body::stream()` now have a stream of `Bytes` instead of `impl Buf` in: https://github.com/aknuds1/warp/blob/bb2f511fc8785784089ebdb12861bbc32e4cef5d/src/filters/body.rs#L77-L78

This following ... some Bytes change I cannot track down, which removed `Buf::to_bytes`.

**Comment by paolobarbolini** [comment]
> This following ... some Bytes change I cannot track down, which removed `Buf::to_bytes`.

It was replaced by [Buf::copy_to_bytes](https://docs.rs/bytes/1.0.1/bytes/buf/trait.Buf.html#method.copy_to_bytes) @koivunej

**Comment by Kiskae** [comment]
It appears 657f1c6 broke the build since `warp::test::handshake` uses `hyper::Client`:

```
   Compiling warp v0.2.5 (https://github.com/aknuds1/warp?branch=chore/upgrade-tokio#bb2f511f)
error[E0433]: failed to resolve: could not find `Client` in `hyper`
   --> /home/x/.cargo/git/checkouts/warp-2008e9c17b453427/bb2f511/src/test.rs:518:36
    |
518 |             let upgrade = ::hyper::Client::builder()
    |                                    ^^^^^^ could not find `Client` in `hyper`

error: aborting due to previous error
```

**Comment by aknuds1** [comment]
> It appears [657f1c6](https://github.com/seanmonstar/warp/commit/657f1c660ce582432ac884e029f6d3349cfac798) broke the build since `warp::test::handshake` uses `hyper::Client`:
> 
> ```
>    Compiling warp v0.2.5 (https://github.com/aknuds1/warp?branch=chore/upgrade-tokio#bb2f511f)
> error[E0433]: failed to resolve: could not find `Client` in `hyper`
>    --> /home/x/.cargo/git/checkouts/warp-2008e9c17b453427/bb2f511/src/test.rs:518:36
>     |
> 518 |             let upgrade = ::hyper::Client::builder()
>     |                                    ^^^^^^ could not find `Client` in `hyper`
> 
> error: aborting due to previous error
> ```

@Kiskae What's the command to reproduce your failure?

**Comment by Kiskae** [comment]
@aknuds1 Calling `cargo build` with the branch as a dependency seems to trigger it:

```
[package]
name = "repro"
version = "0.1.0"
edition = "2018"

[dependencies]
warp = {git = "https://github.com/aknuds1/warp", branch = "chore/upgrade-tokio"}
```

**Comment by aknuds1** [comment]
@Kiskae Aha, I see the problem, it's because src/test.rs contains public utilities and depends on hyper::Client.

**Comment by nickelc** [comment]
> @Kiskae Aha, I see the problem, it's because src/test.rs contains public utilities and depends on hyper::Client. Guess I'll have to reintroduce the dependency on hyper's client feature @seanmonstar.
i think another option could be to make it an optional `test` feature that requires `hyper/client`

**Review by seanmonstar — APPROVED** [review]

**Comment by aknuds1** [comment]
Thanks @seanmonstar! :tada:

**Pull request merged**
